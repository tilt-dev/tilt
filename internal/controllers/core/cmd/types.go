package cmd

import "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"

type Cmd = v1alpha1.Cmd
type CmdSpec = v1alpha1.CmdSpec
type CmdStatus = v1alpha1.CmdStatus
type CmdStateRunning = v1alpha1.CmdStateRunning
type CmdStateTerminated = v1alpha1.CmdStateTerminated

const LabelManifest = v1alpha1.LabelManifest
