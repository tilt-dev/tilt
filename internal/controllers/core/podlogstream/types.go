package podlogstream

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PodLogStream = v1alpha1.PodLogStream
type PodLogStreamSpec = v1alpha1.PodLogStreamSpec
type PodLogStreamStatus = v1alpha1.PodLogStreamStatus
type ContainerLogStreamStatus = v1alpha1.ContainerLogStreamStatus
type ObjectMeta = metav1.ObjectMeta
