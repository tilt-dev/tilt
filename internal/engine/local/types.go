package local

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type Cmd = v1alpha1.Cmd
type CmdList = v1alpha1.CmdList
type CmdStatus = v1alpha1.CmdStatus
type CmdSpec = v1alpha1.CmdSpec
type CmdStateWaiting = v1alpha1.CmdStateWaiting
type CmdStateTerminated = v1alpha1.CmdStateTerminated
type CmdStateRunning = v1alpha1.CmdStateRunning
type ObjectMeta = metav1.ObjectMeta
