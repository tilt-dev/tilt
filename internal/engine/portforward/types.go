package portforward

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PortForward = v1alpha1.PortForward
type PortForwardSpec = v1alpha1.PortForwardSpec
type PortForwardStatus = v1alpha1.PortForwardStatus
type ObjectMeta = metav1.ObjectMeta
type Forward = v1alpha1.Forward
