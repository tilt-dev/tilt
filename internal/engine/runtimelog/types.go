package runtimelog

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PodLogStream = v1alpha1.PodLogStream
type PodLogStreamSpec = v1alpha1.PodLogStreamSpec
type ObjectMeta = metav1.ObjectMeta
