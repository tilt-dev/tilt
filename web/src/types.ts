import { Alert } from "./alerts"
import { Facet } from "./facets"

export enum SocketState {
  Loading,
  Reconnecting,
  Closed,
  Active,
}

export enum ResourceView {
  Log,
  Alerts,
  Facets = 2,
}

export enum TriggerMode {
  TriggerModeAuto,
  TriggerModeManualAfterInitial,
  TriggerModeManualIncludingInitial,
}

// what is the status of the resource in the cluster
export enum RuntimeStatus {
  Ok = "ok",
  Pending = "pending",
  Error = "error",
  NotApplicable = "not_applicable",
}

// What is the status of the resource with respect to Tilt
export enum ResourceStatus {
  Building, // Tilt is actively doing something (e.g., docker build or kubectl apply)
  Pending, // not building, healthy, or unhealthy, but presumably on its way to one of those (e.g., queued to build, or ContainerCreating)
  Healthy, // e.g., build succeeded and pod is running and healthy
  Unhealthy, // e.g., last build failed, or CrashLoopBackOff
  Warning, // e.g., an undismissed restart
  None, // e.g., a manual build that has never executed
}

export type SnapshotHighlight = {
  beginningLogID: string
  endingLogID: string
  text: string
}

export enum ShowFatalErrorModal {
  Default,
  Show,
  Hide,
}

export type Snapshot = {
  view: Proto.webviewView
  isSidebarClosed: boolean
  path?: string
  snapshotHighlight?: SnapshotHighlight | null
}

// A plaintext representation of a line of the log,
// with metadata to render it in isolation.
//
// The metadata should be stored as primitive fields
// so that React's default caching behavior will kick in.
export type LogLine = {
  // We assume that 'text' does not contain a newline
  text: string
  manifestName: string
  level: string
}
