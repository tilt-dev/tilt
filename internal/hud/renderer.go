package hud

import (
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/gdamore/tcell"
	"github.com/windmilleng/tilt/internal/dockercompose"
	"github.com/windmilleng/tilt/internal/hud/view"
	"github.com/windmilleng/tilt/internal/model"
	"github.com/windmilleng/tilt/internal/rty"
)

type Renderer struct {
	rty    rty.RTY
	screen tcell.Screen
	mu     *sync.Mutex
	clock  func() time.Time
}

func NewRenderer(clock func() time.Time) *Renderer {
	return &Renderer{
		mu:    new(sync.Mutex),
		clock: clock,
	}
}

func (r *Renderer) Render(v view.View, vs view.ViewState) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.rty != nil {
		layout := r.layout(v, vs)
		err := r.rty.Render(layout)
		if err != nil {
			return err
		}
	}
	return nil
}

var cText = tcell.Color232
var cLightText = tcell.Color243
var cGood = tcell.ColorGreen
var cBad = tcell.ColorRed
var cPending = tcell.ColorYellow

var statusColors = map[string]tcell.Color{
	"Running":                          cGood,
	"ContainerCreating":                cPending,
	"Pending":                          cPending,
	"Error":                            cBad,
	"CrashLoopBackOff":                 cBad,
	string(dockercompose.StatusInProg): cPending,
	string(dockercompose.StatusUp):     cGood,
	string(dockercompose.StatusDown):   cBad,
	"Completed":                        cGood,
}

func (r *Renderer) layout(v view.View, vs view.ViewState) rty.Component {
	l := rty.NewFlexLayout(rty.DirVert)
	if vs.ShowNarration {
		l.Add(renderNarration(vs.NarrationMessage))
		l.Add(rty.NewLine())
	}

	l.Add(r.renderResourceHeader(v))
	l.Add(r.renderResources(v, vs))
	l.Add(r.renderPaneHeader(vs))
	l.Add(r.renderLogPane(v, vs))
	l.Add(r.renderFooter(v, keyLegend(v, vs)))

	var ret rty.Component = l
	if vs.LogModal.TiltLog == view.TiltLogFullScreen {
		ret = r.renderTiltLog(v, vs, keyLegend(v, vs), ret)
	} else if vs.LogModal.ResourceLogNumber != 0 {
		ret = r.renderResourceLogModal(v.Resources[vs.LogModal.ResourceLogNumber-1], ret)
	}

	ret = r.maybeAddAlertModal(vs, ret)

	return ret
}

func (r *Renderer) maybeAddAlertModal(vs view.ViewState, layout rty.Component) rty.Component {
	if vs.AlertMessage != "" {
		l := rty.NewLines()
		l.Add(rty.TextString(""))

		msg := "   " + vs.AlertMessage + "   "
		l.Add(rty.Fg(rty.TextString(msg), tcell.ColorDefault))
		l.Add(rty.TextString(""))

		b := rty.NewBox(l)
		b.SetTitle("! Alert !")
		layout = r.renderModal(rty.Fg(b, tcell.ColorRed), layout, false)
	}
	return layout
}

func (r *Renderer) renderLogPane(v view.View, vs view.ViewState) rty.Component {
	l := rty.NewConcatLayout(rty.DirHor)
	log := rty.NewTextScrollLayout("log")
	log.Add(rty.TextString(v.Log))
	l.Add(log)
	height := 7
	if vs.LogModal.TiltLog == view.TiltLogMinimized {
		height = 1
	}
	return rty.NewFixedSize(l, rty.GROW, height)
}

func (r *Renderer) renderPaneHeader(vs view.ViewState) rty.Component {
	var s string
	switch vs.LogModal.TiltLog {
	case view.TiltLogFullScreen:
		s = "(l) minimize log ↓"
	case view.TiltLogPane:
		s = "(l) maximize log ↑"
	case view.TiltLogMinimized:
		s = "(l) expand log ↑"
	}
	l := rty.NewLine()
	l.Add(rty.NewFillerString('─'))
	l.Add(rty.TextString(fmt.Sprintf(" %s ", s)))
	l.Add(rty.NewFillerString('─'))
	return l
}

func (r *Renderer) renderStatusBar(v view.View) rty.Component {
	sb := rty.NewStringBuilder()
	sb.Text(" ") // Indent
	errorCount := 0
	for _, res := range v.Resources {
		if isInError(res, v.TriggerMode) {
			errorCount++
		}
	}
	if errorCount == 0 && v.TiltfileErrorMessage == "" {
		sb.Fg(cGood).Text("✓").Fg(tcell.ColorDefault).Fg(cText).Text(" OK").Fg(tcell.ColorDefault)
	} else {
		var errorCountMessage string
		var tiltfileError strings.Builder
		s := "error"
		if errorCount > 1 {
			s = "errors"
		}

		if errorCount > 0 {
			errorCountMessage = fmt.Sprintf(" %d %s", errorCount, s)
		}

		if v.TiltfileErrorMessage != "" {
			_, err := tiltfileError.WriteString(" • Tiltfile error")
			if err != nil {
				// This space intentionally left blank
			}
		}
		sb.Fg(cBad).Text("✖").Fg(tcell.ColorDefault).Fg(cText).Textf("%s%s", errorCountMessage, tiltfileError.String()).Fg(tcell.ColorDefault)
	}
	return rty.Bg(rty.OneLine(sb.Build()), tcell.ColorWhiteSmoke)
}

func (r *Renderer) renderFooter(v view.View, keys string) rty.Component {
	footer := rty.NewConcatLayout(rty.DirVert)
	footer.Add(r.renderStatusBar(v))
	l := rty.NewConcatLayout(rty.DirHor)
	sbRight := rty.NewStringBuilder()
	sbRight.Text(keys)
	l.AddDynamic(rty.NewFillerString(' '))
	l.Add(sbRight.Build())
	footer.Add(l)
	return rty.NewFixedSize(footer, rty.GROW, 2)
}

func keyLegend(v view.View, vs view.ViewState) string {
	defaultKeys := "Browse (↓ ↑), Expand (→) ┊ (enter) log, (b)rowser ┊ (ctrl-C) quit  "
	if vs.LogModal.TiltLog == view.TiltLogFullScreen {
		return "Scroll (↓ ↑) ┊ cycle (l)og view "
	} else if vs.LogModal.ResourceLogNumber != 0 {
		return "Scroll (↓ ↑) ┊ (esc) close logs "
	} else if vs.AlertMessage != "" {
		return "Tilt (l)og ┊ (esc) close alert "
	} else if v.TriggerMode == model.TriggerManual {
		return "Build (space) ┊ " + defaultKeys
	}
	return defaultKeys
}

func isInError(res view.Resource, triggerMode model.TriggerMode) bool {
	return statusColor(res, triggerMode) == cBad
}

func isCrashing(res view.Resource) bool {
	return (res.IsK8S() && res.K8SInfo().PodRestarts > 0) ||
		res.LastBuild().Reason.Has(model.BuildReasonFlagCrash) ||
		res.CurrentBuild.Reason.Has(model.BuildReasonFlagCrash) ||
		res.PendingBuildReason.Has(model.BuildReasonFlagCrash) ||
		res.IsDC() && res.DockerComposeTarget().Status() == string(dockercompose.StatusCrash)
}

func bestLogs(res view.Resource) string {
	// A build is in progress, triggered by an explicit edit.
	if res.CurrentBuild.StartTime.After(res.LastBuild().FinishTime) &&
		!res.CurrentBuild.Reason.IsCrashOnly() {
		return string(res.CurrentBuild.Log)
	}

	// A build is in progress, triggered by a pod crash.
	if res.CurrentBuild.StartTime.After(res.LastBuild().FinishTime) &&
		res.CurrentBuild.Reason.IsCrashOnly() {
		return res.CrashLog + "\n\n" + string(res.CurrentBuild.Log)
	}

	// The last build was an error.
	if res.LastBuild().Error != nil {
		return string(res.LastBuild().Log)
	}

	if k8sInfo, ok := res.ResourceInfo.(view.K8SResourceInfo); ok {
		// Two cases:
		// 1) The last build finished before this pod started
		// 2) This log is from an in-place container update.
		// in either case, prepend them to pod logs.
		if (res.LastBuild().StartTime.Equal(k8sInfo.PodUpdateStartTime) ||
			res.LastBuild().StartTime.Before(k8sInfo.PodCreationTime)) &&
			len(res.LastBuild().Log) > 0 {
			return string(res.LastBuild().Log) + "\n" + res.ResourceInfo.RuntimeLog()
		}

		// The last build finished, but the pod hasn't started yet.
		if res.LastBuild().StartTime.After(k8sInfo.PodCreationTime) {
			return string(res.LastBuild().Log)
		}
	}

	if res.IsTiltfile {
		return string(res.LastBuild().Log)
	}

	return string(res.LastBuild().Log) + "\n" + res.ResourceInfo.RuntimeLog()
}

func (r *Renderer) renderTiltLog(v view.View, vs view.ViewState, keys string, background rty.Component) rty.Component {
	l := rty.NewConcatLayout(rty.DirVert)
	sl := rty.NewTextScrollLayout(logScrollerName)
	l.Add(r.renderPaneHeader(vs))
	sl.Add(rty.TextString(v.Log))
	l.AddDynamic(sl)
	l.Add(r.renderFooter(v, keys))
	return rty.NewModalLayout(background, l, 1, true)
}

func (r *Renderer) renderResourceLogModal(res view.Resource, background rty.Component) rty.Component {
	s := bestLogs(res)
	if len(strings.TrimSpace(s)) == 0 {
		s = fmt.Sprintf("No log output for %s", res.Name)
	}

	l := rty.NewTextScrollLayout(logScrollerName)
	l.Add(rty.TextString(s))
	box := rty.NewGrowingBox()
	box.SetInner(l)
	box.SetTitle(fmt.Sprintf("LOG: %s", res.Name))

	return rty.NewModalLayout(background, box, 0.9, true)
}

func (r *Renderer) renderModal(fg rty.Component, bg rty.Component, fixed bool) rty.Component {
	return rty.NewModalLayout(bg, fg, .9, fixed)
}

func renderNarration(msg string) rty.Component {
	lines := rty.NewLines()
	l := rty.NewLine()
	l.Add(rty.TextString(msg))
	lines.Add(rty.NewLine())
	lines.Add(l)
	lines.Add(rty.NewLine())

	box := rty.Fg(rty.Bg(lines, tcell.ColorLightGrey), cText)
	return rty.NewFixedSize(box, rty.GROW, 3)
}

func (r *Renderer) renderResourceHeader(v view.View) rty.Component {
	l := rty.NewConcatLayout(rty.DirHor)
	l.Add(rty.ColoredString("  RESOURCE NAME ", cLightText))
	l.AddDynamic(rty.NewFillerString(' '))

	k8sCell := rty.ColoredString(" DEPLOY", cLightText)
	l.Add(k8sCell)
	l.Add(middotText())

	buildCell := rty.NewMinLengthLayout(BuildDurCellMinWidth+BuildStatusCellMinWidth, rty.DirHor).
		SetAlign(rty.AlignEnd).
		Add(rty.ColoredString("BUILD STATUS", cLightText))
	l.Add(buildCell)
	l.Add(middotText())
	deployCell := rty.NewMinLengthLayout(DeployCellMinWidth+1, rty.DirHor).
		SetAlign(rty.AlignEnd).
		Add(rty.ColoredString("UPDATED ", cLightText))
	l.Add(deployCell)
	return rty.OneLine(l)
}

func (r *Renderer) renderResources(v view.View, vs view.ViewState) rty.Component {
	rs := v.Resources

	cl := rty.NewConcatLayout(rty.DirVert)

	childNames := make([]string, len(rs))
	for i, r := range rs {
		childNames[i] = r.Name.String()
	}
	// the items added to `l` below must be kept in sync with `childNames` above
	l, selectedResource := r.rty.RegisterElementScroll(resourcesScollerName, childNames)

	if len(rs) > 0 {
		for i, res := range rs {
			l.Add(r.renderResource(res, vs.Resources[i], v.TriggerMode, selectedResource == res.Name.String()))
		}
	}

	cl.Add(l)
	return cl
}

func (r *Renderer) renderResource(res view.Resource, rv view.ResourceViewState, triggerMode model.TriggerMode, selected bool) rty.Component {
	return NewResourceView(res, rv, triggerMode, selected, r.clock).Build()
}

func (r *Renderer) SetUp() (chan tcell.Event, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	screen, err := tcell.NewScreen()
	if err != nil {
		if err == tcell.ErrTermNotFound {
			// The statically-compiled tcell only supports the most common TERM configs.
			// The dynamically-compiled tcell supports more, but has distribution problems.
			// See: https://github.com/gdamore/tcell/issues/252
			term := os.Getenv("TERM")
			return nil, fmt.Errorf("Tilt does not support TERM=%q. "+
				"This is not a common Terminal config. "+
				"If you expect that you're using a common terminal, "+
				"you might have misconfigured $TERM in your .profile.", term)
		}
		return nil, err
	}
	if err = screen.Init(); err != nil {
		return nil, err
	}
	screenEvents := make(chan tcell.Event)
	go func() {
		for {
			screenEvents <- screen.PollEvent()
		}
	}()

	r.rty = rty.NewRTY(screen)

	r.screen = screen

	return screenEvents, nil
}

func (r *Renderer) Reset() {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.screen != nil {
		r.screen.Fini()
	}

	r.screen = nil
}
