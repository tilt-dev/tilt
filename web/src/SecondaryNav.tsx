import React, { PureComponent } from "react"
import styled from "styled-components"
import { Link } from "react-router-dom"
import { LogTrace, LogTraceNav, ResourceView } from "./types"
import * as s from "./style-helpers"

type NavProps = {
  logUrl: string
  traceNav: LogTraceNav | null
  alertsUrl: string
  facetsUrl: string | null
  resourceView: ResourceView
  numberOfAlerts: number
}

let Root = styled.nav`
  display: flex;
  flex-direction: column;
`

let NavList = styled.ul`
  display: flex;
  list-style: none;
  height: ${s.Height.secondaryNav}px;
`

let NavListLower = styled.div`
  display: flex;
  background-color: ${s.Color.grayDark};
  border-top: 2px solid ${s.Color.gray};
  border-bottom: 2px solid ${s.Color.gray};
  height: ${s.Height.secondaryNavLower}px;
  margin-top: ${s.Height.secondaryNavOverlap}px;
  font-size: ${s.FontSize.smallest};
  font-family: ${s.Font.sansSerif};
  align-items: stretch;
`

let NavListLowerItem = styled.div`
  border-right: 2px solid ${s.Color.gray};
  padding: 0 ${s.SizeUnit(0.5)};
  display: flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  text-align: center;

  &:first-child {
    width: ${s.Width.secondaryNavItem}px;
  }
`

let NavListLowerLink = styled(Link)`
  border-right: 2px solid ${s.Color.gray};
  padding: 0 ${s.SizeUnit(0.5)};
  display: flex;
  align-items: center;

  background-color: transparent;
  color: ${s.Color.grayLight};
  box-sizing: border-box;
  transition: color, border-color;
  transition-duration: ${s.AnimDuration.default};
  text-decoration: none;
  cursor: pointer;

  &:hover {
    background-color: ${s.Color.gray};
    color: ${s.Color.blue};
  }

  &.is-disabled {
    color: ${s.Color.gray};
    pointer-events: none;
  }
`

let NavListItem = styled.li`
  display: flex;
  align-items: stretch;
`

let NavLink = styled(Link)`
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: ${s.Font.sansSerif};
  font-size: ${s.FontSize.small};
  text-decoration: none;
  width: ${s.Width.secondaryNavItem}px;
  text-align: center;
  border-top-left-radius: ${s.SizeUnit(0.2)};
  border-top-right-radius: ${s.SizeUnit(0.2)};
  color: ${s.Color.grayLight};
  transition: color ${s.AnimDuration.default} ease;

  &.isSelected {
    color: ${s.Color.white};
    background-color: ${s.Color.grayDark};
    box-shadow: -3px -3px 2px 1px ${s.Color.grayDarkest};
  }

  &:hover {
    color: ${s.Color.blue};
  }
`

let Badge = styled.div`
  font-family: ${s.Font.sansSerif};
  font-size: ${s.FontSize.smallest};
  background-color: ${s.Color.white};
  color: ${s.Color.grayDarkest};
  width: ${s.Width.badge}px;
  height: ${s.Width.badge}px;
  border-radius: ${s.Width.badge}px;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: ${s.SizeUnit(0.3)};
`

class SecondaryNav extends PureComponent<NavProps> {
  renderSecondLevelNav() {
    let traceIsSelected = this.props.resourceView === ResourceView.Trace
    let traceNav = this.props.traceNav
    if (!traceIsSelected || !traceNav) {
      return null
    }

    let current = traceNav.current
    let prev = traceNav.prev
    let next = traceNav.next

    let secondLevelItems = []
    secondLevelItems.push(
      <NavListLowerItem key={"trace-label"}>For Update</NavListLowerItem>
    )

    secondLevelItems.push(
      <NavListLowerLink
        key={"trace-prev"}
        to={prev?.url || ""}
        title="Previous"
        className={prev ? "" : "is-disabled"}
      >
        ◄
      </NavListLowerLink>
    )

    let summary = `${current.index + 1} / ${traceNav.count}`
    secondLevelItems.push(
      <NavListLowerItem key={"trace-summary"}>{summary}</NavListLowerItem>
    )

    secondLevelItems.push(
      <NavListLowerLink
        key={"trace-next"}
        to={next?.url || ""}
        title="Next"
        className={next ? "" : "is-disabled"}
      >
        ►
      </NavListLowerLink>
    )

    return <NavListLower>{secondLevelItems}</NavListLower>
  }

  render() {
    let traceIsSelected = this.props.resourceView === ResourceView.Trace
    let logIsSelected =
      this.props.resourceView === ResourceView.Log || traceIsSelected
    let alertsIsSelected = this.props.resourceView === ResourceView.Alerts
    let facetsIsSelected = this.props.resourceView === ResourceView.Facets

    let secondLevelNav = this.renderSecondLevelNav()

    let facetItem = null
    if (this.props.facetsUrl) {
      facetItem = (
        <NavListItem>
          <NavLink
            className={`tabLink ${facetsIsSelected ? "isSelected" : ""}`}
            to={this.props.facetsUrl}
          >
            Facets
          </NavLink>
        </NavListItem>
      )
    }

    // The number of alerts should be for the selected resource
    return (
      <Root className="secondaryNav">
        <NavList>
          <NavListItem>
            <NavLink
              className={logIsSelected ? "isSelected" : ""}
              to={this.props.logUrl}
            >
              Logs
            </NavLink>
          </NavListItem>
          <NavListItem>
            <NavLink
              className={`secondaryNavLink--alerts ${
                alertsIsSelected ? "isSelected" : ""
              }`}
              to={this.props.alertsUrl}
            >
              Alerts
              {this.props.numberOfAlerts > 0 && (
                <Badge>{this.props.numberOfAlerts}</Badge>
              )}
            </NavLink>
          </NavListItem>
          {facetItem}
        </NavList>
        {secondLevelNav}
      </Root>
    )
  }
}

export default SecondaryNav
