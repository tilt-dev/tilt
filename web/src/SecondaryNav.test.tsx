import React from "react"
import renderer from "react-test-renderer"
import SecondaryNav from "./SecondaryNav"
import { MemoryRouter } from "react-router"
import { ResourceView } from "./types"

it("shows logs", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SecondaryNav
          logUrl="/r/foo"
          alertsUrl="/r/foo/alerts"
          traceNav={null}
          facetsUrl={null}
          resourceView={ResourceView.Log}
          numberOfAlerts={0}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("shows error pane", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SecondaryNav
          logUrl="/r/foo"
          alertsUrl="/r/foo/alerts"
          facetsUrl={null}
          traceNav={null}
          resourceView={ResourceView.Alerts}
          numberOfAlerts={0}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("shows the number of errors in the error tab", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SecondaryNav
          logUrl="/r/foo"
          alertsUrl="/r/foo/alerts"
          facetsUrl={null}
          traceNav={null}
          resourceView={ResourceView.Alerts}
          numberOfAlerts={27}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("shows a facets tab", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <SecondaryNav
          logUrl="/r/foo"
          alertsUrl="/r/foo/alerts"
          facetsUrl="/r/foo/facets"
          traceNav={null}
          resourceView={ResourceView.Facets}
          numberOfAlerts={0}
        />
      </MemoryRouter>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
