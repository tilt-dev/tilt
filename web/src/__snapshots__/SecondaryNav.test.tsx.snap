// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows a facets tab 1`] = `
<nav
  className="sc-bdVaJa bPGLpq secondaryNav"
>
  <ul
    className="sc-bwzfXH iOvBVw"
  >
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE"
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE secondaryNavLink--alerts "
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
      </a>
    </li>
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE tabLink isSelected"
        href="/r/foo/facets"
        onClick={[Function]}
      >
        Facets
      </a>
    </li>
  </ul>
</nav>
`;

exports[`shows error pane 1`] = `
<nav
  className="sc-bdVaJa bPGLpq secondaryNav"
>
  <ul
    className="sc-bwzfXH iOvBVw"
  >
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE"
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE secondaryNavLink--alerts isSelected"
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
      </a>
    </li>
  </ul>
</nav>
`;

exports[`shows logs 1`] = `
<nav
  className="sc-bdVaJa bPGLpq secondaryNav"
>
  <ul
    className="sc-bwzfXH iOvBVw"
  >
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE isSelected"
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE secondaryNavLink--alerts "
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
      </a>
    </li>
  </ul>
</nav>
`;

exports[`shows the number of errors in the error tab 1`] = `
<nav
  className="sc-bdVaJa bPGLpq secondaryNav"
>
  <ul
    className="sc-bwzfXH iOvBVw"
  >
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE"
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li
      className="sc-EHOje dmhADK"
    >
      <a
        className="sc-bZQynM juTImE secondaryNavLink--alerts isSelected"
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
        <div
          className="sc-gzVnrw ewXMFN"
        >
          27
        </div>
      </a>
    </li>
  </ul>
</nav>
`;
